Std/Tactic/Omega/Frontend.lean

```diff
@@ -0,0 +1,464 @@
+Split a disjunction in a `MetaProblem`, and if we find a new usable fact
+call `omegaImpl` in both branches.
+-/
+partial def splitDisjunction (m : MetaProblem) (g : MVarId) : OmegaM Unit := g.withContext do
+  match m.disjunctions with
+    | [] => throwError "omega did not find a contradiction:\n{m.problem}"
+    | h :: t =>
+      trace[omega] "Case splitting on {← inferType h}"
+      let ctx ← getMCtx
+      let (⟨g₁, h₁⟩, ⟨g₂, h₂⟩) ← cases₂ g h
+      trace[omega] "Adding facts:\n{← g₁.withContext <| inferType (.fvar h₁)}"
+      let m₁ := { m with facts := [.fvar h₁], disjunctions := t }
+      let r ← savingState do
+        let (m₁, n) ← g₁.withContext m₁.processFacts
+        if 0 < n then
+          omegaImpl m₁ g₁
+          pure true
+        else
+          pure false
+      if r then
+        trace[omega] "Adding facts:\n{← g₂.withContext <| inferType (.fvar h₂)}"
+        let m₂ := { m with facts := [.fvar h₂], disjunctions := t }
+        omegaImpl m₂ g₂
+      else
+        trace[omega] "No new facts found."
+        setMCtx ctx
+        splitDisjunction { m with disjunctions := t } g
+
+/-- Implementation of the `omega` algorithm, and handling disjunctions. -/
+partial def omegaImpl (m : MetaProblem) (g : MVarId) : OmegaM Unit := g.withContext do
+  let (m, _) ← m.processFacts
+  guard m.facts.isEmpty
+  let p := m.problem
+  trace[omega] "Extracted linear arithmetic problem:\nAtoms: {← atomsList}\n{p}"
+  let p' ← if p.possible then p.elimination else pure p
+  trace[omega] "After elimination:\nAtoms: {← atomsList}\n{p'}"
+  match p'.possible, p'.proveFalse?, p'.proveFalse?_spec with
+  | true, _, _ =>
+    splitDisjunction m g
+  | false, .some prf, _ =>
+    trace[omega] "Justification:\n{p'.explanation?.get}"
+    let prf ← instantiateMVars (← prf)
+    trace[omega] "omega found a contradiction, proving {← inferType prf}"
+    trace[omega] "{prf}"
+    g.assign prf
+
+end
+
+/--
+Given a collection of facts, try prove `False` using the omega algorithm,
+and close the goal using that.
+-/
+def omega (facts : List Expr) (g : MVarId) (cfg : OmegaConfig := {}) : MetaM Unit :=
+  OmegaM.run (omegaImpl { facts } g) cfg
+
+open Lean Elab Tactic Parser.Tactic
+
+/--
+The `omega` tactic, for resolving integer and natural linear arithmetic problems.
+
+It is not yet a full decision procedure (no "dark" or "grey" shadows),
+but should be effective on many problems.
+
+We handle hypotheses of the form `x = y`, `x < y`, `x ≤ y`, and `k ∣ x` for `x y` in `Nat` or `Int`
+(and `k` a literal), along with negations of these statements.
+
+We decompose the sides of the inequalities as linear combinations of atoms.
+
+If we encounter `x / k` or `x % k` for literal integers `k` we introduce new auxiliary variables
+and the relevant inequalities.
+
+On the first pass, we do not perform case splits on natural subtraction.
+If `omega` fails, we recursively perform a case split on
+a natural subtraction appearing in a hypothesis, and try again.
```

Comment:
This is not a user-facing tactic description. It should show more examples and explain what it does from a user perspective.